lib/ddtrace/trace_segment.rb

```diff
@@ -0,0 +1,110 @@
+require 'forwardable'
+require 'ddtrace/ext/distributed'
+require 'ddtrace/ext/priority'
+require 'ddtrace/ext/sampling'
+
+module Datadog
+  # Serializable construct representing a trace
+  class TraceSegment
+    extend Forwardable
+
+    TAG_NAME = 'name'.freeze
+    TAG_RESOURCE = 'resource'.freeze
+    TAG_SERVICE = 'service'.freeze
+
+    SPANS_METHODS = [
+      :any?,
+      :count,
+      :empty?,
+      :length,
+      :size
+    ].freeze
+
+    attr_reader \
+      :id,
+      :spans,
+      :tags
+
+    def initialize(
+      spans,
+      agent_sample_rate: nil,
+      hostname: nil,
+      id: nil,
+      lang: nil,
+      name: nil,
+      origin: nil,
+      process_id: nil,
+      rate_limiter_rate: nil,
+      resource: nil,
+      root_span_id: nil,
+      rule_sample_rate: nil,
+      runtime_id: nil,
+      sample_rate: nil,
+      sampling_priority: nil,
+      service: nil
+    )
+      @id = id
+      @root_span_id = root_span_id
+      @spans = spans || []
+      @tags = {}
+
+      # Set well-known tags
+      self.agent_sample_rate = agent_sample_rate
+      self.hostname = hostname
+      self.lang = lang
+      self.name = name
+      self.origin = origin
+      self.process_id = process_id
+      self.rate_limiter_rate = rate_limiter_rate
+      self.resource = resource
+      self.rule_sample_rate = rule_sample_rate
+      self.runtime_id = runtime_id
+      self.sample_rate = sample_rate
+      self.sampling_priority = sampling_priority
+      self.service = service
+    end
+
+    def_delegators :spans, *SPANS_METHODS
+
+    # Define tag accessors
+    {
+      agent_sample_rate: Ext::Sampling::TAG_AGENT_RATE,
+      hostname: Ext::NET::TAG_HOSTNAME,
+      lang: Ext::Runtime::TAG_LANG,
+      name: TAG_NAME,
+      origin: Ext::DistributedTracing::TAG_ORIGIN,
+      process_id: Ext::Runtime::TAG_PID,
+      rate_limiter_rate: Ext::Sampling::TAG_RATE_LIMITER_RATE,
+      resource: TAG_RESOURCE,
+      rule_sample_rate: Ext::Sampling::TAG_RULE_SAMPLE_RATE,
+      runtime_id: Ext::Runtime::TAG_ID,
+      sample_rate: Ext::Sampling::TAG_SAMPLE_RATE,
+      sampling_priority: Ext::DistributedTracing::TAG_SAMPLING_PRIORITY,
+      service: TAG_SERVICE
+    }.each do |tag_name, tag_key|
+      define_method(tag_name) { tags[tag_key] }
+      define_method(:"#{tag_name}=") { |value| tags[tag_key] = value }
+    end
+
+    def keep!
+      self.sampling_priority = Datadog::Ext::Priority::USER_KEEP
+    end
+
+    def reject!
+      self.sampling_priority = Datadog::Ext::Priority::USER_REJECT
+    end
```

Comment:
As we provide these those two modifier methods, we should have a counterpart: a way to read the sampling state of this trace segment.

Reply:
That should just be `trace.sampling_priority`, or do you mean you want a helper like `sampled?`?

Reply:
`sampled?` is good!